src/hb-ot-layout.cc

```diff
@@ -1046,6 +1064,97 @@ hb_ot_layout_get_size_params (hb_face_t    *face,
+ * @first_param_id: (out) (allow-none): The first ‘name’ table name ID used to specify
+ *                  strings for user-interface labels for the feature
+ *                  parameters. (Must be zero if numParameters is zero.)
+ * @char_count: (out) (allow-none): The count of characters for which this feature
+ *              provides glyph variants. (May be zero.)
+ * @characters: (out) (allow-none): The Unicode Scalar Value of the characters for which
+ *              this feature provides glyph variants.
+ *
+ * Returns bunch of NameIDs and other related information for a tag,
+ * and sets NameIDs parameters (suffixed with _id) to 0 in case
+ * the requested tag not found or the tag doesn't support it.
+ *
+ * Return value: true if could find any feature with the tag, false otherwise
+ *
+ * Since: REPLACEME
+ **/
+hb_bool_t
+hb_ot_layout_get_feature_params (hb_face_t       *face,
+				 hb_tag_t         tag,
+				 unsigned int    *label_id,
+				 unsigned int    *tooltip_id,
+				 unsigned int    *sample_id,
+				 unsigned int    *num_named_parameters,
+				 unsigned int    *first_param_id,
+				 unsigned int    *char_count,
+				 hb_codepoint_t **characters)
+{
+  const OT::FeatureParams &feature_params =
+    _get_gsubgpos_matched_feature_params (face, tag);
+  if (&feature_params != &OT::Null (OT::FeatureParams))
+  {
+    const OT::FeatureParamsStylisticSet& stylistic_set_params =
+      feature_params.get_stylistic_set_params (tag);
+    if (&stylistic_set_params != &OT::Null (OT::FeatureParamsStylisticSet)) /* ssXX */
+    {
+#define PARAM(a, A) if (a) *a = A
+      PARAM(label_id, stylistic_set_params.uiNameID);
+      // ssXX features don't have the rest
+      PARAM(tooltip_id, 0);
+      PARAM(sample_id, 0);
+      PARAM(num_named_parameters, 0);
+      PARAM(first_param_id, 0);
+      PARAM(char_count, 0);
+      PARAM(characters, nullptr);
+      return true;
+    }
+    const OT::FeatureParamsCharacterVariants& character_variants_params =
+      feature_params.get_character_variants_params (tag);
+    if (&character_variants_params != &OT::Null (OT::FeatureParamsCharacterVariants)) /* cvXX */
+    {
+      PARAM(label_id, character_variants_params.featUILableNameID);
+      PARAM(tooltip_id, character_variants_params.featUITooltipTextNameID);
+      PARAM(sample_id, character_variants_params.sampleTextNameID);
+      PARAM(num_named_parameters, character_variants_params.numNamedParameters);
+      PARAM(first_param_id, character_variants_params.firstParamUILabelNameID);
+      PARAM(char_count, character_variants_params.characters.len);
+      unsigned int len = character_variants_params.characters.len;
+      if (len != 0 && characters)
+      {
+	hb_codepoint_t *chars = (hb_codepoint_t*) malloc (len * sizeof (hb_codepoint_t));
+	for (unsigned int i = 0; i < len; ++i)
+	  chars[i] = character_variants_params.characters[i];
+	*characters = chars;
```

Comment:
May be instead of copying, just return a pointer to `character_variants_params.characters`, this way the caller does not need to free this (and note in the documentation that this pointer is only valid as long as the face is valid, so the caller should copy itself).

Reply:
I like that to be honest but it seems this is not the convention as it is done here [here](https://github.com/harfbuzz/harfbuzz/blob/f24b0b9728f87d7599867a4ba6ec7d30da43b869/src/hb-ot-name-table.hh#L100) also apparently but the main problem is it uses [3 bytes characters](https://github.com/harfbuzz/harfbuzz/blob/435b1878e710b67e21872572ad9fd5ed1369e97f/src/hb-ot-layout-common-private.hh#L452) and I guess it would be more nice to resolve it for the client.

Reply:
Also does `Unicode Scalar Value` correspond to hb_codepoint_t? If not, what should be done instead?

Reply:
The alternative is to have the client give us an allocated array to fill, but this makes the API a bit more complex. Allocating internally using `malloc` and expecting the client to call `free` is bad design IMHO, since we might be using different memory allocation methods, especially since we allow overriding them.
 
Unicode Scalar Value is another name for UTF-32 code points, so yes they are the same.

Reply:
I don’t think we use `name` table `get_name()` any where, but if you look closely you will see it doesn’t do any allocation and takes in an allocated buffer and length.

You can do the same here, which is even a better alternative. Allocating internally using `malloc` and expecting the client to call `free` is bad design IMHO, since we might be using different memory allocation methods, especially since we allow overriding them.
 
Unicode Scalar Value is another name for UTF-32 code points, so yes they are the same.

Reply:
> Allocating internally using malloc and expecting the client to call free is bad design IMHO

Oh OK, so will follow that

Reply:
```
hb_ot_layout_get_feature_params (hb_face_t       *face,
				 hb_tag_t         tag,
				 unsigned int    *label_id,
				 unsigned int    *tooltip_id,
				 unsigned int    *sample_id,
				 unsigned int    *num_named_parameters,
				 unsigned int    *first_param_id,
				 unsigned int    *char_count,
				 unsigned int     buffer_length,
				 hb_codepoint_t **characters)
```

Should the signature should look like this?

Reply:
I haven’t seen one in the wild, but you can make one with FontTools, I guess.

Reply:
Too much hassle :D